src/Map.ts

```diff
@@ -543,329 +549,235 @@ export class Map extends maplibregl.Map {
    */
-  setPrimaryLanguage(language: LanguageString = defaults.primaryLanguage) {
+  private setPrimaryLanguage(language: LanguageString | string) {
+    const styleLanguage = this.getStyleLanguage();
+
+    // If the language is set to `STYLE` (which is the SDK default), but the language defined in
+    // the style is `auto`, we need to bypass some verification and modify the languages anyway
+    if (
+      !(
+        language === Language.STYLE &&
+        (styleLanguage === Language.AUTO || styleLanguage === Language.VISITOR)
+      )
+    ) {
+      if (language !== Language.STYLE) {
+        this.languageAlwaysBeenStyle = false;
+      }
+
+      if (this.languageAlwaysBeenStyle) {
+        return;
+      }
+
+      // No need to change the language
+      if (this.primaryLanguage === language && !this.forceLanguageUpdate) {
+        return;
+      }
+    }
+
+    if (!isLanguageSupported(language as string)) {
+      console.warn(`The language "${language}" is not supported.`);
+      return;
+    }
+
     if (this.primaryLanguage === Language.STYLE_LOCK) {
       console.warn(
         "The language cannot be changed because this map has been instantiated with the STYLE_LOCK language flag.",
       );
       return;
     }
 
-    if (!isLanguageSupported(language as string)) {
-      return;
-    }
+    this.primaryLanguage = language as LanguageString;
+    let languageNonStyle: LanguageString = language as LanguageString;
 
-    this.primaryLanguage = language;
+    // STYLE needs to be translated into one of the other language,
+    // this is why it's addressed first
+    if (language === Language.STYLE) {
+      if (!styleLanguage) {
+        console.warn("The style has no default languages.");
+        return;
+      }
 
-    this.onStyleReady(() => {
-      if (language === Language.AUTO) {
-        return this.setPrimaryLanguage(getBrowserLanguage());
+      if (!isLanguageSupported(styleLanguage)) {
+        console.warn("The language defined in the style is not valid.");
+        return;
       }
 
-      const layers = this.getStyle().layers;
+      languageNonStyle = styleLanguage as LanguageString;
+    }
 
-      // detects pattern like "{name:somelanguage}" with loose spacing
-      const strLanguageRegex = /^\s*{\s*name\s*(:\s*(\S*))?\s*}$/;
+    // may be overwritten below
+    let langStr: string | LanguageString = Language.LOCAL;
 
-      // detects pattern like "name:somelanguage" with loose spacing
-      const strLanguageInArrayRegex = /^\s*name\s*(:\s*(\S*))?\s*$/;
+    // will be overwritten below
+    let replacer: Array<any> | string = `{${langStr}}`;
 
-      // for string based bilingual lang such as "{name:latin}  {name:nonlatin}" or "{name:latin}  {name}"
-      const strBilingualRegex =
-        /^\s*{\s*name\s*(:\s*(\S*))?\s*}(\s*){\s*name\s*(:\s*(\S*))?\s*}$/;
+    if (languageNonStyle == Language.VISITOR) {
+      langStr = getBrowserLanguage();
+      replacer = [
```

Comment:
This exact code showed up in the polygon PR.
As said previously, you can pull the `ExpressionSpecification` type from maplibre-gl and define the replacer as `ExpressionSpecification`. This means you can remove the `@ts-ignore`s.

Reply:
Nice!